File: Aula_Ricardo/Programação_orientada_a_objetos/go/Singleton.go
package main

import (
	"fmt"
	"sync"
)

type Configuracao struct {
	Valor string
}

var instance *Configuracao
var once sync.Once

func GetInstance(valor string) *Configuracao {
	once.Do(func() {
		instance = &Configuracao{Valor: valor}
	})
	return instance
}

func main() {
	config1 := GetInstance("Configuração 1")
	fmt.Println(config1.Valor)

	config2 := GetInstance("Configuração 2")
	fmt.Println(config2.Valor)

	fmt.Println(config1 == config2)
}

File: Aula_Ricardo/Programação_orientada_a_objetos/java/Singleton.java
public class Configuracao {
    private static Configuracao instance;
    
    private String valor;

    private Configuracao(String valor) {
        this.valor = valor;
    }

    public static Configuracao getInstance(String valor) {
        if (instance == null) {
            instance = new Configuracao(valor);
        }
        return instance;
    }

    public String getValor() {
        return valor;
    }

    public static void main(String[] args) {
        Configuracao config1 = Configuracao.getInstance("Configuração 1");
        System.out.println(config1.getValor());

        Configuracao config2 = Configuracao.getInstance("Configuração 2");
        System.out.println(config2.getValor());

        System.out.println(config1 == config2);
    }
}
